File: src/services/api/graph.ts
/**
 * Graph API 模块
 * 提供图谱数据、反向链接、重要性排名、知识集群等 API
 */
import { invoke } from "@tauri-apps/api/core";
import type { GraphData, BacklinkInfo, CardImportance, KnowledgeCluster } from "./types";

/**
 * 获取完整图谱数据 (包含布局、PageRank、集群)
 */
export async function getData(): Promise<GraphData> {
  return await invoke<GraphData>("get_graph_data");
}

/**
 * 获取指定卡片的反向链接
 * @param cardId 卡片 ID
 * @returns 链接到该卡片的所有卡片信息
 */
export async function getBacklinks(cardId: string): Promise<BacklinkInfo[]> {
  return await invoke<BacklinkInfo[]>("get_backlinks", { cardId });
}

/**
 * 获取卡片重要性排名 (PageRank)
 * @param limit 返回数量限制，默认 50
 */
export async function getCardImportance(limit?: number): Promise<CardImportance[]> {
  return await invoke<CardImportance[]>("get_card_importance", { limit });
}

/**
 * 获取知识集群 (连通分量)
 */
export async function getKnowledgeClusters(): Promise<KnowledgeCluster[]> {
  return await invoke<KnowledgeCluster[]>("get_knowledge_clusters");
}

/**
 * 获取孤立节点 (知识孤岛)
 * @returns 没有任何连接的卡片 ID 列表
 */
export async function getOrphanNodes(): Promise<string[]> {
  return await invoke<string[]>("get_orphan_nodes");
}

/**
 * 重建图谱索引
 */
export async function rebuild(): Promise<void> {
  return await invoke<void>("rebuild_graph");
}

File: src-tauri/src/commands/graph.rs
//! Graph 相关命令
//! 提供图谱数据、反向链接、重要性排名、知识集群等 API

use crate::graph::{self, BacklinkInfo, CardImportance, GraphData, KnowledgeCluster};
use crate::state::AppState;
use tauri::State;

/// 获取完整图谱数据 (包含布局)
#[tauri::command]
pub async fn get_graph_data(state: State<'_, AppState>) -> Result<GraphData, String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    let cards = services.card.get_all().await.map_err(|e| e.to_string())?;
    // 转换为 CardListItem（graph 模块需要的格式）
    let card_list: Vec<_> = cards.into_iter().map(|c| c.into()).collect();
    Ok(graph::compute_layout(card_list))
}

/// 获取指定卡片的反向链接
#[tauri::command]
pub fn get_backlinks(state: State<AppState>, card_id: String) -> Result<Vec<BacklinkInfo>, String> {
    let graph_engine = state
        .graph_engine
        .lock()
        .unwrap()
        .clone()
        .ok_or("Graph engine not initialized")?;

    Ok(graph_engine.get_backlinks(&card_id))
}

/// 获取卡片重要性排名 (PageRank)
#[tauri::command]
pub fn get_card_importance(
    state: State<AppState>,
    limit: Option<usize>,
) -> Result<Vec<CardImportance>, String> {
    let graph_engine = state
        .graph_engine
        .lock()
        .unwrap()
        .clone()
        .ok_or("Graph engine not initialized")?;

    Ok(graph_engine.get_importance_ranking(limit.unwrap_or(50)))
}

/// 获取知识集群 (连通分量)
#[tauri::command]
pub fn get_knowledge_clusters(state: State<AppState>) -> Result<Vec<KnowledgeCluster>, String> {
    let graph_engine = state
        .graph_engine
        .lock()
        .unwrap()
        .clone()
        .ok_or("Graph engine not initialized")?;

    Ok(graph_engine.get_clusters())
}

/// 获取孤立节点 (知识孤岛)
#[tauri::command]
pub fn get_orphan_nodes(state: State<AppState>) -> Result<Vec<String>, String> {
    let graph_engine = state
        .graph_engine
        .lock()
        .unwrap()
        .clone()
        .ok_or("Graph engine not initialized")?;

    Ok(graph_engine.get_orphan_nodes())
}

/// 重建图谱索引
#[tauri::command]
pub async fn rebuild_graph(state: State<'_, AppState>) -> Result<(), String> {
    let graph_engine = state
        .graph_engine
        .lock()
        .unwrap()
        .clone()
        .ok_or("Graph engine not initialized")?;

    // 从数据库获取所有卡片
    let services = state.get_services().ok_or("Vault not initialized")?;
    let cards = services.card.get_all().await.map_err(|e| e.to_string())?;
    let card_list: Vec<_> = cards.into_iter().map(|c| c.into()).collect();
    
    graph_engine.rebuild_with_cards(card_list);
    Ok(())
}
